storage/innobase/fil/fil0fil.cc

```diff
@@ -4847,17 +4815,15 @@ fil_node_complete_io(fil_node_t* node, const IORequest& type)
 		ut_ad(!srv_read_only_mode
 		      || fsp_is_system_temporary(node->space->id));
 
-		++fil_system->modification_counter;
-
-		node->modification_counter = fil_system->modification_counter;
+		node->needs_flush = true;
 
 		if (fil_buffering_disabled(node->space)) {
 
 			/* We don't need to keep track of unflushed
 			changes as user has explicitly disabled
 			buffering. */
 			ut_ad(!node->space->is_in_unflushed_spaces);
-			node->flush_counter = node->modification_counter;
+			node->needs_flush = false;
 
 		} else if (!node->space->is_in_unflushed_spaces) {
```

Comment:
Do we actually need `node->needs_flush` if we already have `space->is_in_unflushed_spaces`?
Yes, there could be some unnecessary `fsync()` calls for the system tablespace or the temporary tablespace if they consist of multiple files, but by default they consist of a single file.

Reply:
I don't think this flag can be removed.
`of_file_flush()` inside `fil_flush_low()` happens without `fil_system->mutex` protection. Consider a multifile tablespace. While we're flushing one file a simultaneous write will happen and we'll never know which file was written to. And we'll can't decide whether we can or can't unset `is_in_unflushed_spaces`.

But anyways `fsync()` is specific to file and IMO `flush_needed` should be a feature of a specific file, not a whole tablespace. Or we'll can't check whether file was flushed and it's safe to close it.